### PR TITLE
[SimplifyCFG] Increase budget for FoldTwoEntryPHINode() if the branch is unpredictable.

### PR DESCRIPTION
**Pull request by tianqingw**
The `!unpredictable` metadata has been present for a long time, but
it's usage in optimizations is still limited. This patch teaches
`FoldTwoEntryPHINode()` to be more aggressive with an unpredictable
branch to reduce mispredictions.

A TTI interface `getBranchMispredictPenalty()` is added to distinguish
between different hardwares to ensure we don't go too far for simpler
cores. For simplicity, only a naive x86 implementation is included for
the time being.

### CODE REVIEW

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp [hunk 1]
```diff
@@ -3549,8 +3550,10 @@ static bool FoldTwoEntryPHINode(PHINode *PN, const TargetTransformInfo &TTI,
   // that need to be moved to the dominating block.
   SmallPtrSet<Instruction *, 4> AggressiveInsts;
   InstructionCost Cost = 0;
-  InstructionCost Budget =
-      TwoEntryPHINodeFoldingThreshold * TargetTransformInfo::TCC_Basic;
+  unsigned Threshold = TwoEntryPHINodeFoldingThreshold;
+  if (IsUnpredictable)
+    Threshold += TTI.getBranchMispredictPenalty();
```
Comment (williamweixiao) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:3555]: add "LLVM_DEBUG(dbgs() << ..." for debugging?

Comment (tianqingw) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:3555]: Done.

File: llvm/include/llvm/Analysis/TargetTransformInfo.h [hunk 2]
```diff
@@ -419,6 +419,11 @@ class TargetTransformInfo {
   /// this factor, it is very likely to be predicted correctly.
   BranchProbability getPredictableBranchThreshold() const;
 
+  // Returns an integer indicating how aggressive the target wants for
+  // eliminating unpredictable branches. A zero return value means extra
+  // optimization applied to them should be minimal.
+  unsigned getBranchMispredictPenalty() const;
```
Comment (RKSimon) [inline llvm/include/llvm/Analysis/TargetTransformInfo.h:425]: Can we be more specific about how this penalty (and the threshold) is used - should it be comparable to TCK_Latency costs (and should it return InstructrionCost?).

**Review by williamweixiao — APPROVED** [review]
LGTM. Let's wait one or two days to see if there are comments from other reviewers.

File: llvm/test/Transforms/SimplifyCFG/two-entry-phi-fold-unpredictable.ll [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+; CHECK-FOLD-NEXT:    [[DOTFCA_1_INSERT:%.*]] = insertvalue { <2 x float>, <2 x float> } [[DOTFCA_0_INSERT]], <2 x float> [[VELOCITY_SROA_14_0]], 1
+; CHECK-FOLD-NEXT:    ret { <2 x float>, <2 x float> } [[DOTFCA_1_INSERT]]
+;
+entry:
+  %cmp = fcmp fast ogt float %speed, 0x3F747AE140000000
+  br i1 %cmp, label %if.then, label %if.end, !unpredictable !0
+
+if.then:
+  %velocity.sroa.0.0.vec.extract = extractelement <2 x float> %velocity.coerce0, i64 0
+  %mul.i.i.i.i = fmul fast float %velocity.sroa.0.0.vec.extract, %velocity.sroa.0.0.vec.extract
```
Comment (DianQK) [inline llvm/test/Transforms/SimplifyCFG/two-entry-phi-fold-unpredictable.ll:69]: Could you run it with `-passes=strip,instnamer`?

**Review by dianqk — COMMENTED** [review]
I'm wondering if it's possible to remove the `FoldTwoEntryPHINode` altogether and hand it in EarlyIfConversion.

**Comment by DianQK** [comment]
> I'm wondering if it's possible to remove the `FoldTwoEntryPHINode` altogether and hand it in EarlyIfConversion.

I believe this is the right approach. The SimplifyCFG should only perform some trivial transforms from if to select in this context. 

@nikic @dtcxzyw @arsenm might provide additional ideas.

File: llvm/test/Transforms/SimplifyCFG/two-entry-phi-fold-unpredictable.ll (hunk 3 above)
Comment (tianqingw) [reply llvm/test/Transforms/SimplifyCFG/two-entry-phi-fold-unpredictable.ll]: Done.

**Comment by tianqingw** [comment]
> I'm wondering if it's possible to remove the `FoldTwoEntryPHINode` altogether and hand it in EarlyIfConversion.

This is intended to be a minimum viable product for 19.x. That's the reason why it's limited to x86 target for the time being. Moving to EarlyIfConversion can be a next step.

>  The SimplifyCFG should only perform some trivial transforms from if to select in this context.

It is still "transforms from if to select". `FoldTwoEntryPHINode()` does things in a very conservative way, that it'll only fold if every instruction in `IfTrue` and `IfFalse` has no side effect, and can be traced from the operands of PHI. No change is done to these instructions. 

I added a `SpeculateUnpredictables` option to prevent redundant runs.

File: llvm/lib/Target/X86/X86TargetTransformInfo.cpp [hunk 4]
```diff
@@ -6756,3 +6756,8 @@ InstructionCost X86TTIImpl::getScalingFactorCost(Type *Ty, GlobalValue *BaseGV,
     return AM.Scale != 0;
   return -1;
 }
+
+InstructionCost X86TTIImpl::getBranchMispredictPenalty() const {
+  // TODO: Hook MispredictPenalty of SchedMachineModel into this.
+  return 14;
```
Comment (nikic) [inline llvm/lib/Target/X86/X86TargetTransformInfo.cpp:6762]: How did you arrive at the value 14? Why is the TODO a TODO and not implemented in this PR?

**Comment by nikic** [comment]
> > I'm wondering if it's possible to remove the `FoldTwoEntryPHINode` altogether and hand it in EarlyIfConversion.
> 
> This is intended to be a minimum viable product for 19.x. That's the reason why it's limited to x86 target for the time being. Moving to EarlyIfConversion can be a next step.

If I understand correctly, EarlyIfConversion already does this optimization, just on the machine level. Currently, it only allows critical path extension by half the mispredict penalty. It would seem reasonable to allow a larger extension for `!unpredictable` branches. So I don't really understand this idea of "implement it in SimplifyCFG now for 19.x and then move it to EarlyIfConversion later". If you plan to move this to EarlyIfConversion I would rather revert this (including from the release branch) and implement it there directly.

My own 2c on SimplifyCFG vs EarlyIfConversion is that doing it in SimplifyCFG can allow it to benefit from other select optimizations in the middle end, at the cost of losing branch based optimizations. Though the way this was implemented, with this transform only running at the very end of the middle-end pipeline, you wouldn't get the benefit of other middle-end optimizations anyway. If interaction with other middle-end optimizations is the reason for having this in SimplifyCFG, it may make sense to have a phase ordering test to demonstrate them.

**Comment by nikic** [comment]
> I'm wondering if it's possible to remove the `FoldTwoEntryPHINode` altogether and hand it in EarlyIfConversion.

I don't think removing it entirely makes sense or is possible, because InstCombine only recognizes many patterns in select form. So it's important that we convert "small" cases to selects in the middle-end. (I don't think the same is really true for the kinds of huge pattern handled here though.)

**Comment by DianQK** [comment]
Any other comments? :)

**Comment by tianqingw** [comment]
Just a heads-up, I'm working on migrating it to CodeGen, though it's a little more work than expected because some of the infrastructures are not there yet (#102101).

**Pull request merged**